### PR TITLE
CFE-208/master: Improve logs related to insufficiently specified process_count body

### PR DESCRIPTION
**Pull request by nickanderson**

### CODE REVIEW

File: cf-agent/verify_processes.c [hunk 1]
```diff
@@ -96,21 +96,21 @@ static bool ProcessSanityChecks(const Attributes *a, const Promise *pp)
     {
         Log(LOG_LEVEL_ERR, "Process select constraint body promised no result (check body definition)");
         PromiseRef(LOG_LEVEL_ERR, pp);
-        return false;
+        ret = false;
```
Comment (olehermanse) [inline cf-agent/verify_processes.c]: This does not fix a memory leak?

File: cf-agent/verify_processes.c [hunk 2]
```diff
@@ -94,9 +96,24 @@ static bool ProcessSanityChecks(const Attributes *a, const Promise *pp)
...
+    if ((a->haveprocess_count) && ((!a->process_count.min_range) && (!a->process_count.max_range)))
+    {
+        Log(LOG_LEVEL_ERR, "process_count used without match_range, this promise will not define a class based on a count of matching promises.");
+        PromiseRef(LOG_LEVEL_ERR, pp);
+        ret = false;
+    }
+
+    if ((a->haveprocess_count) && ((!a->process_count.in_range_define) || (!a->process_count.out_of_range_define)))
+    {
+        Log(LOG_LEVEL_ERR, "process_count body is insufficiently defined. process_count must specify at least one of in_range_define or out_of_range_define. ");
```
Comment (olehermanse) [inline cf-agent/verify_processes.c]: It seems like MPF does this incorrectly then? 

```
Q: ".../cf-agent" --de":    error: Promise belongs to bundle 'cfe_internal_limit_robot_agents_reap' in file '/var/cfengine/inputs/cfe_internal/core/limit_robot_agents.cf' near line 39
Q: ".../cf-agent" --de":    error: Comment is 'Check cf-monitord process if exceed the **number'**
```

This change causes many error messages in the valgrind / bootstrap test. See Travis logs for more details. Please fix these in MPF before merging, if you believe the error message is correct.

Comment (nickanderson) [reply cf-agent/verify_processes.c]: Oh, this is bad logic on my part. I want to error when neither in_range_define nor out_of_range_define is specified.  Basically, in order to have a useful process_count body, you must have match_range and you must have one of either out_of_range_define or in_range_define.


Thanks

File: cf-agent/verify_processes.c (hunk 1 above)
Comment (nickanderson) [reply cf-agent/verify_processes.c]: ¯\_(ツ)_/¯

Comment (nickanderson) [reply cf-agent/verify_processes.c]: I am not sure why this one originally returned early, but non of the others did, so I changed it to be like the others.

**Comment by nickanderson** [comment]
Hummm, this is failing the ago.cf example in core.

```
error: process_count used without match_range, this promise will not define a class based on a count of matching promises.
```

But that is not true, match_range is specified, but its a range of 0.

```
body process_count anyprocs

{
      match_range => "0,0";
      out_of_range_define => { "any_procs" };
}
```

My condition must be bad.

```
    if ((a->haveprocess_count) && ((!a->process_count.min_range) && (!a->process_count.max_range)))
    {
        Log(LOG_LEVEL_ERR, "process_count used without match_range, this promise will not define a class based on a count of matching promises.");
        PromiseRef(LOG_LEVEL_ERR, pp);
        ret = false;
    }
```

Originally, I tried to check a->process_count.match_range, but it said it didn't exist. That's when I saw and switched to process_count.min_range and max_range.

**Comment by nickanderson** [comment]
After some schooling from Vratislav and Craig, I see that an undefined match_range will result in min_range = 0 and max_range = 0. There is no way to differentate between an underfined match_value and explicitly "0,0".

This helps to shed some light on the relationship between the existing conditons related to restart_class being combined with a process_count body or promising explicitly zero processes.

it makes me question this error case:

```
    if (promised_zero && (a->restart_class))
    {
        // restart_class defines a class when there are no matching processes, the expectation is that another promise will use the presence of the class defined to guard a promise that will result in the process running.
        Log(LOG_LEVEL_ERR, "Promise constraint conflicts - '%s' processes cannot have zero count if restarted",
            pp->promiser);
        PromiseRef(LOG_LEVEL_ERR, pp);
        ret = false;
    }
```
promisedzero is defined when min_range and max_range are 0 (or undefined).

```
    bool promised_zero = ((a->process_count.min_range == 0) && (a->process_count.max_range == 0));
```

So, wouldn't that error message happen any time restart_class is used without a process_count body or when the process_count body didn't specify match_value, or when the process_count body specified 0,0 for match_range?

**Comment by nickanderson** [comment]
squashed.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4005/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4005/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4005/

**Comment by olehermanse** [comment]
@cf-bottom one more time in jenkins, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4015/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4015/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4015/

**Pull request merged**